src/main/kotlin/org/rust/ide/refactoring/generateConstructor/GenerateConstructorHandler.kt

```diff
@@ -31,52 +34,85 @@ class GenerateConstructorAction : CodeInsightAction() {
+        generateConstructorBody(context, editor)
     }
 
-    private fun generateConstructorBody(structItem: RsStructItem, editor: Editor) {
+    private fun generateConstructorBody(context: Context, editor: Editor) {
         checkWriteAccessNotAllowed()
-        val chosenFields = showConstructorArgumentsChooser(structItem.project, structItem) ?: return
+        val substitution = context.implBlock?.typeReference?.type?.typeParameterValues ?: emptySubstitution
+        val chosenFields = showConstructorArgumentsChooser(context.struct.project, context.struct, substitution) ?: return
         runWriteAction {
-            insertNewConstructor(structItem, chosenFields, editor)
+            insertNewConstructor(context.struct, context.implBlock, chosenFields, substitution, editor)
         }
     }
 
-    private fun insertNewConstructor(structItem: RsStructItem, selectedFields: List<ConstructorArgument>, editor: Editor) {
+    private fun isImplBlockValid(impl: RsImplItem): Boolean {
+        if (impl.traitRef != null) return false
+
+        return impl.members?.functionList?.find { it.name == "new" } == null
+    }
+
+    private fun insertNewConstructor(structItem: RsStructItem,
+                                     implBlock: RsImplItem?,
+                                     selectedFields: List<ConstructorArgument>,
+                                     substitution: Substitution,
+                                     editor: Editor) {
         checkWriteAccessAllowed()
         val project = editor.project ?: return
         val structName = structItem.name ?: return
         val psiFactory = RsPsiFactory(project)
-        val impl = psiFactory.createInherentImplItem(structName, structItem.typeParameterList, structItem.whereClause)
+        val impl = if (implBlock == null) {
+            val impl = psiFactory.createInherentImplItem(structName, structItem.typeParameterList, structItem.whereClause)
+            structItem.parent.addAfter(impl, structItem) as RsImplItem
+        } else {
+            implBlock
+        }
+
         val anchor = impl.lastChild.lastChild
-        val constructor = getFunction(structItem, selectedFields, psiFactory)
+        val constructor = getFunction(structItem, selectedFields, psiFactory, substitution)
         impl.lastChild.addBefore(constructor, anchor)
-        val insertedImpl = structItem.parent.addAfter(impl, structItem) as RsImplItem
-        editor.caretModel.moveToOffset(insertedImpl.textOffset + insertedImpl.textLength - 1)
+        editor.caretModel.moveToOffset(impl.textOffset + impl.textLength - 1)
     }
 
-    private fun getFunction(structItem: RsStructItem, selectedFields: List<ConstructorArgument>, psiFactory: RsPsiFactory): RsFunction {
+    private fun getFunction(structItem: RsStructItem,
+                            selectedFields: List<ConstructorArgument>,
+                            psiFactory: RsPsiFactory,
+                            substitution: Substitution): RsFunction {
```

Comment:
It should look like the following:

```suggestion
    private fun getFunction(
        structItem: RsStructItem,
        selectedFields: List<ConstructorArgument>,
        psiFactory: RsPsiFactory,
        substitution: Substitution
    ): RsFunction {
```
according to kotlin [style guide](https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting)

`Put parameter on one line` and `Put parameter on separate lines` intentions may help you.
Also, don't forget to check all changed function signature

Reply:
Aah, this intention was what I was looking for! Thanks.